### ISSUE TITLE
Kobo Forma support

### ISSUE THREAD

**Issue by patrick-tang** [koreader/koreader#4291]
* KOReader version:
koreader-kobo-arm-linux-gnueabihf-v2015.11-stable
(launch with KFMon-v1.2.6-1-g9ab7d3a)
* Device:
Kobo Forma
(firmware version: N782890001425,4.1.15,4.11.11911,4.1.15,4.1.15,00000000-0000-0000-0000-000000000377)

#### Issue
In new released Kobo Forma, when launching Koreader, it crashed right away, reporting "unrecognized Kobo model frost".

Complete detailed log:
[crash.log](https://github.com/koreader/koreader/files/2523197/crash.log)

#### Steps to reproduce
After installed, try to launch Koread, it will crash with above crash log.

##### `crash.log` (if applicable)
./luajit: frontend/device/kobo/device.lua:177: unrecognized Kobo model frost
stack traceback:
	[C]: in function 'error'
	frontend/device/kobo/device.lua:177: in main chunk
	[C]: in function 'probeDevice'
	frontend/device.lua:37: in main chunk
	[C]: in function 'require'
	frontend/ui/uimanager.lua:1: in main chunk
	[C]: in function 'require'
	./reader.lua:92: in main chunk
	[C]: at 0x0000bea5

**Comment by Frenzie** [comment]
In frontend/device/kobo/device.lua, you could try adding a definition:

Here's an example:

https://github.com/koreader/koreader/blob/c01d105bb31df51850ce67194842c841682ccf7e/frontend/device/kobo/device.lua#L185-L199

Sorry for the brevity, but I have to be off now. :-)

**Comment by NiLuJe** [comment]
When a Kobo asks you to re-login, that wipes the DB, IIRC. Which means you may need to reboot it for it to process stuff again, provided it hasn't done a bonus factory reset, which would explain 3., as that reinstalls the factory FW, but not 2., because I *think* that wipes the user partition, too.

If it actually did a factory reset, tthat's a fairly visible process, though, so you'd have noticed the process happening on screen.

On the other hand, if that was a simple reboot and it just asked you to re-login just after that, that was simply a trashed database, which can happen (or be misdetected as "useful", even when it isn't), after bad crashes.

Sidebar: starting with the extremely old "stable" release was also a fairly bad idea ;).

**Comment by NiLuJe** [comment]
In the meantime, if you got another crash.log from the hard crash, that might be useful, too ;).

**Comment by patrick-tang** [comment]
@NiLuJe it seems a simple reboot, but something weired because other than a home without any book I also see the settings missing "accounts" submenu.  I forced to restart again, everything is back, including colletion settings.

It was my bad. I forgot to metion when I made the above metioned changes, I actually used the latest beta version and copied the code for daylight ( which I blieve is Kobo Aura One) for my Kobo Forma.  So seems those same code for Aura One not working for Kobo Forma.

What else information I can collect? Would like to fix this with help from you guys.

**Comment by NiLuJe** [comment]
Might have been a botched Nickel restart after a KOReader crash then...

The crash.log from that might help, to begin with ;).

But it's probably going to take someone with shell access to get this right, because I'm expecting the gyro/rotation handling to be wonderfully annoying to deal with.

**Comment by NiLuJe** [comment]
The One is an older device, you'll probably have less terrible results if you start from a Clara, that's what's gotten the most testing out of that generation of devices ;).

**Comment by patrick-tang** [comment]
@NiLuJe
Thanks for the tips. I have copied the setting for Clara, now I'm able to launch Koreader.
However I cannot tap menu items properly. It seems the viewport setting was not correct.
I tried to set up with the resolution, but Koreader then couldn't start.

How can I find out correct viewport setting for this device?
Attached is the log after I add viewport setting (Note: I tried different combination, so there are different log sections).
[crash.log](https://github.com/koreader/koreader/files/2523762/crash.log)

**Comment by Frenzie** [comment]
@patrick-tang The viewport setting is just for KOReader to draw further from the edge of the actual panel, like on the H2O where ~10-11 px are covered up by the bezel.

The full framebuffer will always be the same size, in your case 1920x1440. The viewport is a smaller area within that framebuffer.

The simplest means of quickly testing is to start from scratch on touch workarounds. Getting rid of that `touch_snow_protocol = true,` for example.

If the touch is mirrored in some way, set mirrored to false:

https://github.com/koreader/koreader/blob/c01d105bb31df51850ce67194842c841682ccf7e/frontend/device/kobo/device.lua#L29-L32

What do you mean by "cannot tap menu items properly"? Is there any response?

**Comment by NiLuJe** [comment]
According to the latest news from the KSM thread, you might want to try phoenix + probe_ev ;)

**Comment by patrick-tang** [comment]
@Frenzie 
Sorry I didn't describe clearly. For "cannot tap menu items properly", I meant when I tried to tap the menu items I couldn't activate a menu item by tap on the item, but I had to tap another item (or need to shift the position) - however not like mirrored.
Question:
what is "touch_snow_protocol = true" for? or function of any such touch_xxxx_protocol ?

@NiLuJe 
Do you have the link for KSM thread about "phoenix + probe_ev"? Or a keyword I can search?
I think I tried "touch_probe_ev_epoch_time = true, touch_phoenix_protocol = true" and suspect those settings reboot the device to special state (like another version of firmware).

**Comment by Frenzie** [comment]
So touch is working, just off by a few pixels? That's very weird, but also very easy to work around. Potentially even with viewport instead of event adjustment. Sorry, on my phone atm.

But please try Niluje's suggestion first.

Hold works too?

**Comment by NiLuJe** [comment]
And, nope, my bad, skimmed stuff too rapidly, snow is indeed probably the right protocol, to begin with.

Might need some extra shenanigans, like `touch_probe_ev_epoch_time = true`, or unlikelier, but who knows, `touch_mirrored_x = false`.

Stuff *might* be affected by the rotation you were in before starting KOReader, so I'd keep it simple: Portrait, buttons on the right (or the left?).

**Comment by patrick-tang** [comment]
@NiLuJe 
what is "touch_probe_ev_epoch_time = true" for? 
I don't see Clara using it.

Just want to confirm that hold/swipes work (but buttons don't work). Other problems I notice with current status:
1) the **backlight will keep on when in sleep**.
    If I exit from Koreader and go back to Kobo Home, the backlight will be off in sleep.
    Is this related to backlight class not found issue in error log? 

2) **buttons don't work**, so can only use swipe/tap to turn pages

For above two problems, what should I look at?

**Comment by Frenzie** [comment]
Buttons are really simple, but you need to switch to debug mode, ideally combined with using Telnet or SSH. Isn't there a guide about that on Mobileread somewhere?

Then you simply get output like this:
```
10/29/18-20:03:36 DEBUG key event => type: 1, code: 13(J), value: 1, time: 1540839816.344482
10/29/18-20:03:36 DEBUG AutoSuspend: onInputEvent
10/29/18-20:03:36 DEBUG key event => type: 1, code: 13(J), value: 0, time: 1540839816.426518
```

tl;dr Buttons are the easiest problem to solve. :-)

Unfortunately the closest thing to a basic debugging for dummies guide we've currently got is the one for Android for which I supplied most of the materials, but that one presupposes you already know the basics.

**Comment by NiLuJe** [comment]
Not really, as there's a bunch of different ways you can actually get in.

In this specific instance, using the built-in one (i.e., enablind Nickel's debug features, which enables telnet) is probably a no-go, since Nickel won't be up ;).

That leaves a telnet/SSH package.
I've got [one](https://www.mobileread.com/forums/showthread.php?t=254214), which the potential caveat that it bundles a lot of CLI utilities, too.

@frostchutz has [one](https://github.com/frostschutz/Kobo/tree/master/TelnetFtp), too, which you'll just have to package yourself.

And there's the good old and [venerable](https://www.mobileread.com/forums/showpost.php?p=2681553&postcount=3) package from KevinShort.

Among others, but that'd be my top three ;).

**Comment by Frenzie** [comment]
I think the last one there might be the one I first used to start hacking my H2O.

**Comment by NiLuJe** [comment]
NO. NOPE. NOPE NOPE NOPE.

**Comment by NiLuJe** [comment]
That's why I put mine & frostchutz's first: we go out of our way to actively not mess with any default system files.

**Comment by frostschutz** [comment]
Actually, mine totally does modify the inittab :-) and still uses the hwclock hack too. For other mods I just use udev rule to get things started, but I think I made the telnet one before I learned how to do that... oi should update it, I guess

**Comment by NiLuJe** [comment]
@frostschutz : heretic! :D

**Comment by pazos** [comment]
@NiLuJe inittab does survive firmware upgrades, as far as I remember. /etc/init.d/on-animator.sh does not survive.

I remember when I was using a script launched from inittab to restore on-animator.sh from on-animator-fmon.sh each time I updated the FW and worked.

**Comment by NiLuJe** [comment]
I vaguely recall seeing it pop up in *some* FW updates, but that was a loooong while ago. Or something :D.

**Comment by patrick-tang** [comment]
@Frenzie

You're right about mirrored tapping another day. The tapping was kind of mirrored: it was actually due to the inconsistence of display and tapping. The display followed screen rotation (rotation status before launching Koreader) but tapping was still expecting coordinations of menu items from non-rotated state.

What's your suggestion of the fix?
Currently I have to rotate back before launching Koreader.

**Comment by patrick-tang** [comment]
@Frenzie 
I can telnet into my Forma now. 
Would you provide some simple guide for enabling debug? I would generate some key event log for the button issue.

**Comment by patrick-tang** [comment]
For the light issue, I got following information. It seems there are two files for the light control:
- mxc_msp430.0: for the brightness only
- tlc5947_bl: for color and brightness

ls -lah /sys/class/backlight/
`lrwxrwxrwx    1 root     root           0 Oct 30 22:31 mxc_msp430.0 -> ../../devices/platform/ntx_bl/backlight/mxc_msp430.0
lrwxrwxrwx    1 root     root           0 Oct 30 22:31 tlc5947_bl -> ../../devices/platform/soc/2000000.aips-bus/2000000.spba-bus/2008000.ecspi/spi_master/spi0/spi0.0/backlight/tlc5947_bl`

`[root@(none) ~]# ls -alh /sys/class/backlight/mxc_msp430.0/
total 0
drwxr-xr-x    3 root     root           0 Oct 30 22:31 .
drwxr-xr-x    3 root     root           0 Oct 30 22:31 ..
-r--r--r--    1 root     root        4.0K Oct 30 22:43 actual_brightness
-rw-r--r--    1 root     root        4.0K Oct 30 22:43 bl_power
-rw-r--r--    1 root     root        4.0K Oct 30 22:43 brightness
lrwxrwxrwx    1 root     root           0 Oct 30 22:43 device -> ../../../ntx_bl
-r--r--r--    1 root     root        4.0K Oct 30 22:43 max_brightness
drwxr-xr-x    2 root     root           0 Oct 30 22:43 power
lrwxrwxrwx    1 root     root           0 Oct 30 22:43 subsystem -> ../../../../../class/backlight
-r--r--r--    1 root     root        4.0K Oct 30 22:43 type
-rw-r--r--    1 root     root        4.0K Oct 30 22:31 uevent
`
`[root@(none) ~]# ls -alh /sys/class/backlight/tlc5947_bl/
total 0
drwxr-xr-x    3 root     root           0 Oct 30 22:31 .
drwxr-xr-x    3 root     root           0 Oct 30 22:31 ..
-r--r--r--    1 root     root        4.0K Oct 30 22:43 actual_brightness
-rw-r--r--    1 root     root        4.0K Oct 30 22:43 bl_power
-rw-r--r--    1 root     root        4.0K Oct 30 22:43 brightness
-rw-r--r--    1 root     root        4.0K Oct 30 22:31 color
lrwxrwxrwx    1 root     root           0 Oct 30 22:43 device -> ../../../spi0.0
-rw-r--r--    1 root     root        4.0K Oct 30 22:43 fl_r_en
-r--r--r--    1 root     root        4.0K Oct 30 22:43 max_brightness
-rw-r--r--    1 root     root        4.0K Oct 30 22:31 max_color
-rw-r--r--    1 root     root        4.0K Oct 30 22:43 out
drwxr-xr-x    2 root     root           0 Oct 30 22:43 power
lrwxrwxrwx    1 root     root           0 Oct 30 22:43 subsystem -> ../../../../../../../../../../../class/backlight
-r--r--r--    1 root     root        4.0K Oct 30 22:43 type
-rw-r--r--    1 root     root        4.0K Oct 30 22:31 uevent
`

**Comment by NiLuJe** [comment]
As for enabling debug, you just need to add the `-d` flag in `koreader.sh`, i.e., change [this line](https://github.com/koreader/koreader/blob/f3ce1c224d838bde31d32bbba79e5f11238fba5f/platform/kobo/koreader.sh#L154) to ` ./reader.lua -d "${args}" >>crash.log 2>&1`

**Comment by patrick-tang** [comment]
@NiLuJe 
Let me enable debug now.

For the light, I actually tried to use following, and there is no more "Could not open file" error. However, the light never turns on now (no matter in sleep or not)
```
frontlight_white = "/sys/class/backlight/tlc5947_bl",
frontlight_red = "/sys/class/backlight/tlc5947_bl",
frontlight_green = "/sys/class/backlight/tlc5947_bl",
```

**Comment by patrick-tang** [comment]
@NiLuJe 

Here is what I got for the input. 
ls -lash /dev/input/by-path/
```
     0 drwxr-xr-x    2 root     root          80 Oct 30 23:34 .
     0 drwxr-xr-x    3 root     root         140 Oct 30 23:34 ..
     0 lrwxrwxrwx    1 root     root           9 Oct 30 23:34 platform-1-0024-event -> ../event1
     0 lrwxrwxrwx    1 root     root           9 Oct 30 23:34 platform-ntx_event0-event -> ../event0
```
ls -lash /dev/input/by-path/platform-1-0024-event 
```
     0 lrwxrwxrwx    1 root     root           9 Oct 30 23:34 /dev/input/by-path/platform-1-0024-event -> ../event1
```
ls -lash /dev/input/by-path/platform-ntx_event0-event 
```
     0 lrwxrwxrwx    1 root     root           9 Oct 30 23:34 /dev/input/by-path/platform-ntx_event0-event -> ../event0
```

**Comment by NiLuJe** [comment]
I'm only vaguely familiar with FL handling on Kobo, but this looks wildly different than on previous devices, so I wouldn't expect things to work just by changing path.

**Comment by NiLuJe** [comment]
I'm fairly sure I had a much more readable command to try somewhere to list the inputs, but I'm not on my main box right now, and it's 5AM, so my brain is pretty much gone...

I'll probably have some better ideas tomorrow, sorry :/.

**Comment by patrick-tang** [comment]
@Frenzie @NiLuJe 
I got a debug log file. Can you point me which part I should look at for button issue?
[crash.log](https://github.com/koreader/koreader/files/2532574/crash.log)

I press **next** button for three times and **previous** button for three times. But I see more key event logs than expected.

**Comment by patrick-tang** [comment]
It seems following are the key event for the two buttons: "type 1 code 193" and "type 1 code 194".
How can I associate the event to functions?

```
10/30/18-23:48:23 DEBUG key event => type: 1, code: 193(nil), value: 1, time: 1540957703.88866
10/30/18-23:48:23 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1540957703.88866
10/30/18-23:48:23 DEBUG key event => type: 1, code: 193(nil), value: 0, time: 1540957703.248987
10/30/18-23:48:23 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1540957703.248987
10/30/18-23:48:23 DEBUG BackgroundRunner: _execute() @  1540957703
10/30/18-23:48:23 DEBUG BackgroundRunnerWidget: start running @  1540957703
10/30/18-23:48:24 DEBUG key event => type: 1, code: 194(nil), value: 1, time: 1540957704.958840
10/30/18-23:48:24 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1540957704.958840
10/30/18-23:48:25 DEBUG key event => type: 1, code: 194(nil), value: 0, time: 1540957705.88851
10/30/18-23:48:25 DEBUG input event => type: 0, code: 0(nil), value: 0, time: 1540957705.88851
```

**Comment by NiLuJe** [comment]
Ah, dug up what I was looking for :).

`cat /proc/bus/input/devices` should be much more detailed, and with a bit of luck, some of those key events are actually from the gyro...

Note, that, besides adding the right codes to the map, you'll also need `hasKeys = yes,` in your frost device block.

And I'm still very interested in the fbink tests, to see if we even have a way of dealing with the rotation on launch, like we do on the Oasis...

**Comment by NiLuJe** [comment]
And while we're there, `ntx_hwconfig -v -s /dev/mmcblk0 DisplayPanel` might be useful in devising the native rotation, too, although I'm playing the long game with that one...

**Comment by baskerville** [comment]
> For the light, I actually tried to use following, and there is no more "Could not open file" error. However, the light never turns on now (no matter in sleep or not)
> 
> ```
> frontlight_white = "/sys/class/backlight/tlc5947_bl",
> frontlight_red = "/sys/class/backlight/tlc5947_bl",
> frontlight_green = "/sys/class/backlight/tlc5947_bl",
> ```

I'm afraid this won't work well, because, for example, if the *red* value is 0, the current interface will write 0 to `bl_power` but if the *white* value is greater than 0, `bl_power` should be 31 (or maybe something else on your device)…

In other words, the current frontlight interface assumes that all the `frontlight_*` directories are distinct.

**Comment by baskerville** [comment]
> Do you know if there is any command related to light or OS document I can look at?

While Nickel (Kobo's UI) is running, the output of `lsof | grep '/sys/'` might be helpful.

**Comment by pazos** [comment]
@patrick-tang:
You can also watch for udev events with this program: 
[eventwatcher.zip](https://github.com/koreader/koreader/files/2536327/eventwatcher.zip)

Download it to your kobo, unzip it somewhere using telnet/ssh and, while nickel is running, start the program with `./eventwatcher &` and just modify light values (turn on/off and change front/natural light values). Then you can read events from /root/event.log and paste them here. Hopefully you'll see all light changes you'll need to support the Forma

**Comment by patrick-tang** [comment]
@Frenzie 
Thanks for the help. Now the keys are working. So one of three issues is resolved:
1) keys not working (fixed)
2) lights not working
3) inconsistency of menu display and tapping.

Now I need to look at the suggestions from @baskerville @NiLuJe @pazos for the other two problems.

**Comment by patrick-tang** [comment]
@poire-z 
I tried to use the eventwather you provided, but only got one line of event when I tried to change light in nickel. What I did:
1) change the brightness from 1% -> 100%, then 100%->0%, then 0%->1%
2) turn off "natural light", then turn on
```
/devices/platform/soc/2100000.aips-bus/21a8000.i2c/i2c-2/2-0032/ricoh619-battery/power_supply/mc13892_bat
```

**Comment by patrick-tang** [comment]
@NiLuJe 
Here is the output of command: cat /proc/bus/input/devices
What should I look at for the rotation issue?

```
[root@(none) ~]# cat /proc/bus/input/devices
I: Bus=0019 Vendor=0001 Product=0001 Version=0100
N: Name="gpio-keys"
P: Phys=gpio-keys/input0
S: Sysfs=/devices/platform/ntx_event0/input/input0
U: Uniq=
H: Handlers=event0 
B: PROP=0
B: EV=100013
B: KEY=6 0 0 100000 0 8000000 0
B: MSC=8

I: Bus=0000 Vendor=0000 Product=0000 Version=0000
N: Name="cyttsp5_mt"
P: Phys=1-0024/input0
S: Sysfs=/devices/platform/soc/2100000.aips-bus/21a4000.i2c/i2c-1/1-0024/input/input1
U: Uniq=
H: Handlers=event1 
B: PROP=2
B: EV=f
B: KEY=421 0 0 0 0 0 0 0 0 0 0
B: REL=0
B: ABS=ef30000 0

I: Bus=0018 Vendor=0000 Product=0000 Version=0000
N: Name="FreescaleAccelerometer"
P: Phys=
S: Sysfs=/devices/virtual/input/input2
U: Uniq=mma8652
H: Handlers=event2 
B: PROP=0
B: EV=9
B: ABS=7
```

**Comment by NiLuJe** [comment]
Okay, so at least we'll have access to the accelerometer, although I'm not sure if anyone ever played with it, so, yaaaay.

ntx_hwconfig returning static info is perfectly normal, don't worry ;).

That leaves checking how Nickel itself actually handles rotation, and for that, you'll have to install FBInk to check what it has to say, as explained in [this post](https://www.mobileread.com/forums/showpost.php?p=3764768&postcount=36).

You'll probably want to use a standalone binary from [here](https://www.mobileread.com/forums/showpost.php?p=3766113&postcount=41) to save you a reboot ;).

**Comment by NiLuJe** [comment]
Depending on the results, we might at least be able to honor the correct rotation at startup.

**Comment by pazos** [comment]
@patrick-tang: does your device has /dev/ntx_io?

**Comment by pazos** [comment]
> My Kobo Forma has ntx_io. What is that for?

It can be used to set frontlight (white) value. Is in fact used in kobos without natural light. You can try with `hasNaturalLight = no, `. Does the frontlight works? 

It should work. 

If so try to write a value between 0 and 100 to the color mixer, like `echo 30 >/sys/devices/platform/backlight/tlc5947_bl/color`

**Comment by patrick-tang** [comment]
@pazos 
I was wrong about the result of change I made earlier. It was actually working, no matter I turn on or off the naturalLight.
So the light setting actually turns on light in reading and turns off light in sleep.
But the light adjustment is not synch-ed with Nickel because:
1) In Koreader, changing to different light environment the frontlight doesn't change accordingly like non-Koreader
2) When exiting from Koreader to Nickel I can see obvious light adjustment. And in Nickel frontlight will change accordingly with environment and quite sensitive.

**Comment by NiLuJe** [comment]
Okay, they did rotation right, *joy*.

It'll put the rotation code through its paces (and, err, reports seem to indicate it isn't in tip top shape to begin with), but at least we know we have a way of honoring the startup rotation.

**Comment by NiLuJe** [comment]
@patrick-tang : Hmm, quick question regarding the FBInk test: was the actual "Hi" message always shown properly, as far as rotation is concerned?

**Comment by NiLuJe** [comment]
As far as the light is concerned, it's a vertical swipe on the left edge of the screen, IIRC.

**Comment by patrick-tang** [comment]
@NiLuJe
Does your tool use some kind of system to get the current rotation code? Which part of Koreader to use rotation code for ui and tapping?
Would check and test it in my device.

**Comment by patrick-tang** [comment]
Then the gesture of light adjustmen is same as Nickel. I'll test it.

**Comment by NiLuJe** [comment]
We could get the same info from base/ffi/framebuffer_linux.lua, but we currently don't, because we usually always have a higher-level method to find the UI rotation on the few devices where it matter (i.e., lipc calls on Kindle Oasis).

**Comment by NiLuJe** [comment]
What we do with that info depends on whether my "Hi" was always printed in the right place or not ;).

If it was, we potentially only have touch input to rotate, if it wasn't, everything.

**Comment by patrick-tang** [comment]
@NiLuJe 
I don't see. Need to enable anything?

**Comment by NiLuJe** [comment]
Nope, debug mode is the most verbose we get ;).

**Comment by cramoisi** [comment]
This battery drain reborn itself at each new kobo device. It's just silly.

**Comment by patrick-tang** [comment]
@cramoisi 
Do you know how that was fixed for other models?

**Comment by cramoisi** [comment]
@patrick-tang : the result is the same (the battery is draining) but it was several distinct issues. For my KA1, switching from KSM to KFMon as a launcher did the trick. What about the crash log ? did it say something during the time the device was "sleeping" ? what about the battery log ?

Do you have a sleepcover ? do you put your kobo near something which has a magnet in it ? My KA1 battery doesn't like my 3DS or my phone...

**Comment by Frenzie** [comment]
@patrick-tang With great difficulty.

But yeah, just by looking at the debug logs. Maybe stick a `dmesg` in there for some more info, that kind of thing.

**Comment by NiLuJe** [comment]
One obvious culprit would be accelerometer doing silly things in this case...

**Comment by patrick-tang** [comment]
The real time monitoring for Forma is a little bit difficult because it will drop the wireless connection automatically (and will reconnect if needed in cases like accessing the store. for power saving purpose?)

What I see regarding of power issue for Forma:
1) It is only to Koreader, so if I remembered to quit Koreader after reading, the battery will not be drained overnight;
2) I use KFMon not KSM, so the trick mentioned by @cramoisi would not improve for this case
3) I have sleepcover, but I use power button to put the device to sleep. And if I exit Koreader before sleep, even if there are cell phones around, I don't see battery draining issue.

I don't see battery log all the time. Let me try more times.

**Comment by shuoroser** [comment]
Hi all, I have Kobo Forma 32Gb , add Kute Monitoring File 1.2.8 . Please guide me lauch Koreader on Forma. Thanks

**Comment by NiLuJe** [comment]
Okay, new plan to deal with the initial rotation:

Could anyone with shell access check that, under Nickel, with the device in landscape mode, doing `echo 1 > /sys/class/graphics/fb0/rotate` properly puts it back in Portrait with buttons on the right?

(If it doesn't, try with other numbers, counting down from 3 until 0).

**Comment by patrick-tang** [comment]
@NiLuJe 
the command has no effect, no matter what value I put there.

**Comment by NiLuJe** [comment]
Eh, that might be enough for us once Nickel's gone, thanks!

I'll post a test koreader.sh script (much) later tonight.

**Comment by NiLuJe** [comment]
There, check if [this version](https://github.com/NiLuJe/koreader/blob/b1c697a76ac0722f055508339d7665c8b993fe24/platform/kobo/koreader.sh) of the koreader.sh startup script allows you to start properly, no matter the orientation.

**Comment by pickx** [comment]
I'm not sure if warm light control is supposed to work with this initial support, but it doesn't on my device. Both sliders seem to be controlling regular light.

**Comment by pazos** [comment]
@pickx 

> I'm not sure if warm light control is supposed to work with this initial support, but it doesn't on my device. Both sliders seem to be controlling regular light.

@NiLuJe added support based on this thread feedback and @patrick-tang said that it worked for him.

If you have shell access you can try https://github.com/koreader/koreader/issues/4291#issuecomment-434912117. Also please post the output of
`udevadm info --attribute-walk --path=/sys/class/backlight/tlc5947_bl/`

Current sysfs interface write on [bl_power & brightness](https://github.com/koreader/koreader/blob/master/frontend/device/sysfs_light.lua#L95-L104). Newer devices may need to write to color instead.

**Comment by pickx** [comment]
@pazos 
I did some experimentation with tlc5947_bl. I'm a newbie, bear with me if all of this is obvious.

`/sys/class/backlight/tlc5947_bl/` indeed controls the natural light color on my Forma. Writing to bl_power or brightness does not seem like the right approach because Nickel does not seem to write to these values. Going back and forth from Nickel to koreader and checking the attributes of `tlc5947_bl`, it seems Nickel instead just uses the color attribute, as you have said. 

Nickel doesn't allow fine-grained control over natural light (at least on Forma), instead the slider has 11 steps going from 10 to 0, which gives 10 color presets (10 is whitest, 0 reddest), and sets the "out" attribute via some combination with the backlight brightness (I'm guessing color presets are just scalars).

Fine-grained control can probably be achieved by modifiying the "out" attribute directly, though I'm not sure such functionality is wanted. Keeping sync with Nickel seems like a more reasonable approach. I personally don't find such fine-grained control useful.

**Comment by baskerville** [comment]
As I said earlier, the output of `lsof | grep '/sys/'` could be helpful.

**Comment by pazos** [comment]
@pickx: for the record, on a BQ Cervantes 4:

```
ls -lah  /sys/class/backlight/
total 0
drwxr-xr-x  2 root root 0 Nov 18 15:35 .
drwxr-xr-x 32 root root 0 Nov 18 15:34 ..
lrwxrwxrwx  1 root root 0 Nov 18 15:37 lm3630a_led -> ../../devices/platform/imx-i2c.0/i2c-0/0-0036/backlight/lm3630a_led
lrwxrwxrwx  1 root root 0 Nov 18 15:37 lm3630a_leda -> ../../devices/platform/imx-i2c.0/i2c-0/0-0036/backlight/lm3630a_leda
lrwxrwxrwx  1 root root 0 Nov 18 15:37 lm3630a_ledb -> ../../devices/platform/imx-i2c.0/i2c-0/0-0036/backlight/lm3630a_ledb
lrwxrwxrwx  1 root root 0 Nov 18 15:37 mxc_msp430_fl.0 -> ../../devices/platform/mxc_msp430_fl.0/backlight/mxc_msp430_fl.0
```

This device has both the "classic" interface (being lm3630a_ledb for white leds and lm3630a_leda for red leds) and the newer color mixer interface (lm3630a_led is the color mixer and  mxc_msp430_fl.0 is for controlling the brightness). It seems that the Kobo Forma doesn't have the "classic" interface. Other recent models, like the Clara, happen to have both interfaces (and the classic is used https://github.com/koreader/koreader/blob/master/frontend/device/kobo/device.lua#L192-L198)

## Color mixer

```
udevadm info --attribute-walk --path=/sys/class/backlight/lm3630a_led
  looking at device '/devices/platform/imx-i2c.0/i2c-0/0-0036/backlight/lm3630a_led':
    KERNEL=="lm3630a_led"
    SUBSYSTEM=="backlight"
    DRIVER==""
    ATTR{bl_power}=="31"
    ATTR{brightness}=="50"
    ATTR{max_brightness}=="100"
    ATTR{type}=="raw"
    ATTR{color}=="1"
    ATTR{max_color}=="10"
    ATTR{ramp}=="0"
    ATTR{ramp_on}=="0"
    ATTR{ramp_off}=="0"
    ATTR{ramp_up}=="0"
    ATTR{ramp_down}=="0"

  looking at parent device '/devices/platform/imx-i2c.0/i2c-0/0-0036':
    KERNELS=="0-0036"
    SUBSYSTEMS=="i2c"
    DRIVERS=="lm3630a_bl"
    ATTRS{name}=="lm3630a_bl"
    ATTRS{modalias}=="i2c:lm3630a_bl"
    ATTRS{microamps_requested_vdd_fl_lm36}=="0"

  looking at parent device '/devices/platform/imx-i2c.0/i2c-0':
    KERNELS=="i2c-0"
    SUBSYSTEMS=="i2c"
    DRIVERS==""
    ATTRS{name}=="imx-i2c"

  looking at parent device '/devices/platform/imx-i2c.0':
    KERNELS=="imx-i2c.0"
    SUBSYSTEMS=="platform"
    DRIVERS=="imx-i2c"
    ATTRS{modalias}=="platform:imx-i2c"

  looking at parent device '/devices/platform':
    KERNELS=="platform"
    SUBSYSTEMS==""
    DRIVERS==""
```

bl_power and brightness attributes of the color mixer are taken from mxc_msp430_fl.0 and you can not write directly to lm3630a_led/bl_power or lm3630a_led/brightness.

## In short
With the frontlight/naturalLight turned off the following commands will turn on both with maximum values

```
# turn on frontlight
echo 31 > /sys/class/backlight/mxc_msp430_fl.0/bl_power
echo 100 > /sys/class/backlight/mxc_msp430_fl.0/brightness
# set color mix (0 - 10)
echo 0 > /sys/class/backlight/lm3630a_led/color
```

**Comment by patrick-tang** [comment]
@NiLuJe 
I've create a PR as the key mapping is wrong for Forma. (https://github.com/koreader/koreader/pull/4376)

And also the screen rotation handling is still not working properly in Forma:
The book and gesture following the rotation but the menu not, so still having problem.

**Comment by pazos** [comment]
@patrick-tang: Could you please give us feedback about natural(orange) light? Does it works? 

About key mapping: in nickel key mapping depends on orientation, right?

**Comment by patrick-tang** [comment]
@pazos 
The natural light doesn't work in koreader.

The key mapping follows orientation, meaning the two keys will switch accordingly with orientation.

**Comment by cairnsh** [comment]
This replacement for the function `setNaturalBrightness` in `frontend/device/sysfs_light.lua` seems to work okay for me on the Forma that I got:

``function SysfsLight:setNaturalBrightness(brightness, warmth)
    if not brightness then
        brightness = self.current_brightness
    end
    if not warmth then
        warmth = self.current_warmth
    end

    warmth = math.floor(warmth/10)

    if brightness > 0 then
        self:_write_value("/sys/class/backlight/mxc_msp430.0/brightness", brightness)
        self:_write_value("/sys/class/backlight/tlc5947_bl/color", 10 - warmth)
    else
        self:_write_value("/sys/class/backlight/mxc_msp430.0/brightness", 0)
    end

    self.current_brightness = brightness
    self.current_warmth = warmth
end```

It seems to be unnecessary to set `bl_power`.

**Comment by pazos** [comment]
@cairnsh: good work. Did you test if sync with nickel still works?

I guess we can add a new var named frontlight_mixer in https://github.com/koreader/koreader/blob/master/frontend/device/sysfs_light.lua#L11 and check if current device has frontlight_mixer with

```
if self.frontlight_mixer then
    warmth = math.floor(warmth/10)
    self:_write_value(self.frontlight_white, brightness)
    self:_write_value(self.frontlight_mixer, 10 - warmth)
else
    ...
end
```

**Comment by pazos** [comment]
This diff against master should work for the forma. The same aproach should work for other devices too 

```
diff --git a/frontend/device/kobo/device.lua b/frontend/device/kobo/device.lua
index fd19a7cd..0512dc19 100644
--- a/frontend/device/kobo/device.lua
+++ b/frontend/device/kobo/device.lua
@@ -215,7 +215,7 @@ local KoboFrost = Kobo:new{
     hasNaturalLight = yes,
     frontlight_settings = {
         frontlight_white = "/sys/class/backlight/mxc_msp430.0",
-        frontlight_red = "/sys/class/backlight/tlc5947_bl",
+        frontlight_mixer = "/sys/class/backlight/tlc5947_bl",
     }
 }
 
diff --git a/frontend/device/sysfs_light.lua b/frontend/device/sysfs_light.lua
index 0252741e..cdb40559 100644
--- a/frontend/device/sysfs_light.lua
+++ b/frontend/device/sysfs_light.lua
@@ -9,6 +9,7 @@ local SysfsLight = {
     frontlight_white = nil,
     frontlight_red = nil,
     frontlight_green = nil,
+    frontlight_mixer = nil,
     current_brightness = 0,
     current_warmth = 0,
     white_gain = 25,
@@ -56,29 +57,36 @@ function SysfsLight:setNaturalBrightness(brightness, warmth)
         warmth = self.current_warmth
     end
 
-    local red = 0
-    local green = 0
-    local white = 0
-    if brightness > 0 then
-        -- On Nickel, the values for white/red/green are roughly linearly dependent
-        -- on the 4th root of brightness and warmth.
-        white = math.min(self.white_gain * math.pow(brightness, self.exponent) *
+    -- Newer devices use a mixer instead of writting values per color.
+    if self.frontlight_mixer then
+        warmth = math.floor(warmth/10)
+        self:_write_value(self.frontlight_white, brightness)
+        self:_write_value(self.frontlight_mixer, 10 - warmth)
+    else
+        local red = 0
+        local green = 0
+        local white = 0
+        if brightness > 0 then
+            -- On Nickel, the values for white/red/green are roughly linearly dependent
+            -- on the 4th root of brightness and warmth.
+            white = math.min(self.white_gain * math.pow(brightness, self.exponent) *
                              math.pow(100 - warmth, self.exponent) + self.white_offset, 255)
-    end
-    if warmth > 0 then
-        red = math.min(self.red_gain * math.pow(brightness, self.exponent) *
+        end
+        if warmth > 0 then
+            red = math.min(self.red_gain * math.pow(brightness, self.exponent) *
                            math.pow(warmth, self.exponent) + self.red_offset, 255)
-        green = math.min(self.green_gain * math.pow(brightness, self.exponent) *
+            green = math.min(self.green_gain * math.pow(brightness, self.exponent) *
                              math.pow(warmth, self.exponent) + self.green_offset, 255)
-    end
+        end
 
-    white = math.max(white, 0)
-    red = math.max(red, 0)
-    green = math.max(green, 0)
+        white = math.max(white, 0)
+        red = math.max(red, 0)
+        green = math.max(green, 0)
 
-    self:_set_light_value(self.frontlight_white, math.floor(white))
-    self:_set_light_value(self.frontlight_green, math.floor(green))
-    self:_set_light_value(self.frontlight_red, math.floor(red))
+        self:_set_light_value(self.frontlight_white, math.floor(white))
+        self:_set_light_value(self.frontlight_green, math.floor(green))
+        self:_set_light_value(self.frontlight_red, math.floor(red))
+    end
 
     self.current_brightness = brightness
     self.current_warmth = warmth
```

**Comment by Frenzie** [comment]
You get syntax highlighting with
````
```diff
````
or lua/python/bash/etc. depending on what's in there ;-)

The diff looks good.

**Comment by NiLuJe** [comment]
I *should* be getting my hands on a Forma tomorrow, so I'll be able to join in on the fun ;).

**Comment by NiLuJe** [comment]
On Nickel's side, ColorSetting is still [1500 ... 6400], so I'll have to check how the normalization fares, because it's now only 10 steps on Nickel's side, while we're doing 100 in `KoboPowerD:_syncKoboLightOnStart` ;).

(Also, brightness seems to be written to actual_brightness (0-100) @ mxc_msp430.0, but I haven't checked what our code does on that front).

**Comment by NiLuJe** [comment]
And before anyone asks: yes, that weird band of light near the LED is real, and either a pretty serious design defect, or Kobo's QA team is entirely composed of a group of blind monkeys.
